File: backend/app/schemas/ai.py
"""TestPilot AI — AI / LLM Pydantic Schemas."""

from __future__ import annotations

from pydantic import Field

from app.schemas.common import BaseSchema


class ChatMessage(BaseSchema):
    """A single message in a chat conversation."""

    role: str = Field(..., description="'user' or 'assistant'")
    content: str


class ChatRequest(BaseSchema):
    """Request to chat with the codebase."""

    repository_id: str
    messages: list[ChatMessage]
    context_files: list[str] = Field(
        default_factory=list,
        description="Specific files to include in context",
    )
    max_tokens: int = Field(default=2048, le=8192)


class ChatResponse(BaseSchema):
    """Response from codebase chat."""

    message: ChatMessage
    sources: list[str] = Field(default_factory=list, description="Source file paths referenced")
    tokens_used: int


class CodeSearchRequest(BaseSchema):
    """Semantic code search request."""

    repository_id: str
    query: str = Field(..., min_length=3, description="Natural language search query")
    limit: int = Field(default=10, ge=1, le=50)
    search_type: str = Field(
        default="hybrid",
        description="Search strategy: 'semantic' | 'structural' | 'hybrid'",
    )
    filters: dict[str, str] = Field(
        default_factory=dict,
        description="Optional filters: language, file_type, etc.",
    )


class CodeSearchResult(BaseSchema):
    """A single code search result."""

    file_path: str
    language: str
    snippet: str
    score: float
    function_name: str | None
    class_name: str | None
    line_start: int | None
    line_end: int | None


class CodeSearchResponse(BaseSchema):
    """Response from code search."""

    results: list[CodeSearchResult]
    total: int
    query: str
    search_type: str


class ImpactAnalysisRequest(BaseSchema):
    """Request for manual impact analysis."""

    repository_id: str
    changed_files: list[str]
    depth: int = Field(default=3, ge=1, le=10, description="Graph traversal depth")


class RiskScoreRequest(BaseSchema):
    """Request for PR risk scoring."""

    pr_id: str
    include_historical: bool = Field(
        default=True,
        description="Include historical bug pattern analysis",
    )

File: frontend/src/lib/api/ai.ts
import { client } from "./client";

export interface ChatMessage {
  role: "user" | "assistant";
  content: string;
}

export interface ChatResponse {
  message: ChatMessage;
  sources: string[];
  tokens_used: number;
}

export interface CodeSearchResult {
  file_path: string;
  language: string;
  snippet: string;
  score: number;
  function_name: string | null;
  class_name: string | null;
  line_start: number | null;
  line_end: number | null;
}

export interface CodeSearchResponse {
  results: CodeSearchResult[];
  total: number;
  query: string;
  search_type: string;
}

export interface ImpactAnalysisResponse {
  changed_files: string[];
  affected_modules: string[];
  impact_radius: number;
  per_file: Record<string, string[]>;
}

export const aiApi = {
  chat: async (repositoryId: string, messages: ChatMessage[], contextFiles: string[] = []) => {
    const res = await client.post<ChatResponse>("/ai/chat", {
      repository_id: repositoryId,
      messages,
      context_files: contextFiles,
    });
    return res.data;
  },

  search: async (repositoryId: string, query: string, limit = 10, searchType = "hybrid") => {
    const res = await client.post<{ success: boolean; data: CodeSearchResponse }>("/ai/search", {
      repository_id: repositoryId,
      query,
      limit,
      search_type: searchType,
    });
    return res.data.data;
  },

  runImpactAnalysis: async (repositoryId: string, changedFiles: string[], depth = 3) => {
    const res = await client.post<{ success: boolean; data: ImpactAnalysisResponse }>("/ai/impact-analysis", {
      repository_id: repositoryId,
      changed_files: changedFiles,
      depth,
    });
    return res.data.data;
  },

  getRiskScore: async (prId: string, includeHistorical = true) => {
    const res = await client.post<{
      success: boolean;
      data: {
        pr_id: string;
        risk_level: "low" | "medium" | "high" | "critical";
        risk_score: number;
        risk_factors: string[];
        analysis_status: string;
      };
    }>("/ai/risk-score", {
      pr_id: prId,
      include_historical: includeHistorical,
    });
    return res.data.data;
  },
};
